src/platform/w3dengine/client/gui/gadget/w3dprogressbar.cpp

```diff
@@ -13,17 +13,192 @@
+                g_theWindowManager->Win_Draw_Line(0xFFFFFFFF, 1.0f, start_x, start_y, end_x, start_y);
+                g_theWindowManager->Win_Draw_Line(0xFFC8C8C8, 1.0f, start_x, start_y, start_x, end_y);
+            }
+        }
+    }
 }
 
 void W3D_Gadget_Progress_Bar_Image_Draw(GameWindow *progress_bar, WinInstanceData *data)
 {
-#ifdef GAME_DLL
-    Call_Function<void, GameWindow *, WinInstanceData *>(0x007CB520, progress_bar, data);
+    unsigned int progress = 0;
+#ifdef GAME_DLL // temporary since we can't change the definition of Win_Send_System_Msg at this point and we can't cast a
+                // pointer to an unsigned int on 64 bit
+    progress = reinterpret_cast<unsigned int>(progress_bar->Win_Get_User_Data());
 #endif
+
+    int screen_x;
+    int screen_y;
+    int width;
+    int height;
+    progress_bar->Win_Get_Screen_Position(&screen_x, &screen_y);
+    progress_bar->Win_Get_Size(&width, &height);
+
+    int x = data->m_imageOffset.x;
+    int y = data->m_imageOffset.y;
+    const Image *left_end_image;
+    const Image *right_end_image;
+    const Image *bar_right_end_image;
+    const Image *repeating_center_image;
+    const Image *bar_repeating_center_image;
+
+    if ((progress_bar->Win_Get_Status() & WIN_STATUS_ENABLED) != 0) {
+        if ((data->m_state & 2) != 0) {
+            left_end_image = progress_bar->Win_Get_Instance_Data()->m_hiliteDrawData[0].image;
+            right_end_image = progress_bar->Win_Get_Instance_Data()->m_hiliteDrawData[1].image;
+            bar_right_end_image = progress_bar->Win_Get_Instance_Data()->m_hiliteDrawData[5].image;
+            repeating_center_image = progress_bar->Win_Get_Instance_Data()->m_hiliteDrawData[2].image;
+            bar_repeating_center_image = progress_bar->Win_Get_Instance_Data()->m_hiliteDrawData[6].image;
+        } else {
+            left_end_image = progress_bar->Win_Get_Instance_Data()->m_enabledDrawData[0].image;
+            right_end_image = progress_bar->Win_Get_Instance_Data()->m_enabledDrawData[1].image;
+            bar_right_end_image = progress_bar->Win_Get_Instance_Data()->m_enabledDrawData[5].image;
+            repeating_center_image = progress_bar->Win_Get_Instance_Data()->m_enabledDrawData[2].image;
+            bar_repeating_center_image = progress_bar->Win_Get_Instance_Data()->m_enabledDrawData[6].image;
+        }
+    } else {
+        left_end_image = progress_bar->Win_Get_Instance_Data()->m_disabledDrawData[0].image;
+        right_end_image = progress_bar->Win_Get_Instance_Data()->m_disabledDrawData[1].image;
+        bar_right_end_image = progress_bar->Win_Get_Instance_Data()->m_disabledDrawData[5].image;
+        repeating_center_image = progress_bar->Win_Get_Instance_Data()->m_disabledDrawData[2].image;
+        bar_repeating_center_image = progress_bar->Win_Get_Instance_Data()->m_disabledDrawData[6].image;
+    }
+
+    if (left_end_image != nullptr && right_end_image != nullptr && repeating_center_image != nullptr
```

Comment:
test for `bar_repeating_center_image` missing here.

Reply:
Original did not have this check.